securedrop_client/queue.py

```diff
@@ -0,0 +1,235 @@
+from sqlalchemy.orm import scoped_session
+from sqlalchemy.orm.session import Session
+from typing import Any, Union, Optional, Type, Tuple
+
+from securedrop_client import storage
+from securedrop_client.crypto import GpgHelper, CryptoError
+from securedrop_client.db import File, Message
+
+
+logger = logging.getLogger(__name__)
+
+
+class ApiInaccessibleError(Exception):
+
+    def __init__(self, message: Optional[str] = None) -> None:
+        if not message:
+            message = ('API is inaccessible either because there is no client or because the '
+                       'client is not properly authenticated.')
+        super().__init__(message)
+
+
+class ApiJob(QObject):
+
+    '''
+    Signal that is emitted after an job finishes successfully.
+    '''
+    success_signal = pyqtSignal('PyQt_PyObject')
+
+    '''
+    Signal that is emitted if there is a failure during the job.
+    '''
+    failure_signal = pyqtSignal(Exception)
+
+    def __init__(self) -> None:
+        super().__init__(None)  # `None` because the QOjbect has no parent
+
+    def _do_call_api(self, api_client: API, session: Session) -> None:
+        if not api_client:
+            raise ApiInaccessibleError()
+
+        try:
+            result = self.call_api(api_client, session)
+        except AuthError as e:
+            raise ApiInaccessibleError() from e
+        except RequestTimeoutError:
+            logger.debug('Job {} timed out'.format(self))
+            raise
+        except Exception as e:
+            logger.error('Job {} raised an exception: {}: {}'
+                         .format(self, type(e).__name__, e))
+            self.failure_signal.emit(e)
+        else:
+            self.success_signal.emit(result)
+
+    def call_api(self, api_client: API, session: Session) -> Any:
+        '''
+        Method for making the actual API call and handling the result.
+
+        This MUST resturn a value if the API call and other tasks were successful and MUST raise
+        an exception if and only iff the tasks failed. Presence of a raise exception indicates a
+        failure.
+        '''
+        raise NotImplementedError
+
+
+class DownloadSubmissionJob(ApiJob):
+
+    CHUNK_SIZE = 4096
+
+    def __init__(
+        self,
+        submission_type: Union[Type[File], Type[Message]],
+        submission_uuid: str,
+        data_dir: str,
+        gpg: GpgHelper,
+    ) -> None:
+        super().__init__()
+        self.data_dir = data_dir
+        self.submission_type = submission_type
+        self.submission_uuid = submission_uuid
+        self.gpg = gpg
+
+    def call_api(self, api_client: API, session: Session) -> Any:
+        db_object = session.query(self.submission_type) \
+            .filter_by(uuid=self.submission_uuid).one()
+
+        etag, download_path = self._make_call(db_object, api_client)
+
+        if not self._check_file_integrity(etag, download_path):
+            raise RuntimeError('Downloaded file had an invalid checksum.')
```

Comment:
we need a followup issue to handle this `RuntimeError`, likely behavior should be to enqueue the download job again as the file was corrupted during download

Reply:
I would also need to delete the file on disk and update the database.